Virtualize swift compiler outputs

cachemeifyoucan opened the pull request:
<!-- What's in this pull request? -->
Step 1 of making swift compiler output cacheable. Virtualize swift compiler outputs (not everything yet, but most of the important output to show feasibility) and add a mechanics to verify the output determinism through the newly added function.

Need some changes in LLVM https://github.com/apple/llvm-project/pull/6112

Feedbacks about how to structurally pass through VirtualBackends in swift compiler is welcomed.

<!--
If this pull request resolves any GitHub issues, link them.
For information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

cachemeifyoucan committed:
Virtualize swift-frontend outputs

Using a virutal output backend to capture all the outputs from
swift-frontend invocation. This allows future virtualization of swift
compiler outputs.

cachemeifyoucan committed:
Add an option to verify the deterministic output of swift compiler

Add option to test swift compiler determinism by running the job twice
and compare the output produced.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci Please test

cachemeifyoucan commented:
Ah, I break lldb of course. let me try smoke test first.

@swift-ci please smoke test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

artemcm reviewed (dismissed):
Looks good to me.
Cool to have the frontend output determinism tests in-suite.

cachemeifyoucan commented:
For deterministic check to be really useful, need to add an option to bypass the laziness when modhash matches.

Also thinking I should just put the output backend into `ASTContext` which will make the code a lot cleaner.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
Update: fix the tests and remove withOutputFile helper function to encourage use of output backend.

Now properly capture swiftdeps output but it is NOT deterministic.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan committed:
Add an option to force compiling object file output

Add an option to skip module hash checking that checks if swift compiler can
skip object file generation because it is going to produce the same
result. Always generation object file output when flag is used so we can
check for output determinism.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

benlangmuir commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -123,37 +128,37 @@ emitLoadedModuleTraceForAllPrimariesIfNeeded(ModuleDecl *mainModule,
-  return os;
+/// Diagnose output file error to ASTContext.
+void diagnoseOutputError(ASTContext &Ctx, StringRef OutputFilename,
+                         StringRef ErrorMsg) {
```
Nit: how about take `Error` and transform to string internally?

cachemeifyoucan commented on include/swift/AST/ASTContext.h:
```diff
@@ -1450,6 +1456,18 @@ class ASTContext final {
 
+  /// Get the output backend. The output backend needs to be initialized via
+  /// constructor or `setOutputBackend`.
+  llvm::vfs::OutputBackend &getOutputBackend() const {
```
You can see an earlier version that adds OutputBackend to lots of function interfaces: https://github.com/apple/swift/commit/b96c260056c6c01a5f84917aceb2a76fe503c633#diff-7a5c341be974c445676d32e0607d37377a7d789212f4fb0e058e3bd092d1ae89

I don't see other good place to pass output backend through swift compiler without putting it in the function signatures everywhere.

benlangmuir replied:
Okay, seems like there isn't a better alternative.

cachemeifyoucan commented on lib/Frontend/FrontendOptions.cpp:
```diff
@@ -173,6 +173,48 @@ bool FrontendOptions::doesActionRequireInputs(ActionType action) {
   llvm_unreachable("Unknown ActionType");
 }
 
+bool FrontendOptions::doesActionRequireOutputBackend(ActionType action) {
```
Probably not. I was switching between if it should be `setupOutputBackedIfNeeded` and make this check or skip this check. I guess I made the wrong choice of mixing them. I will remove this check and not use `IfNeeded` in setup function.

cachemeifyoucan committed:
Support deterministic checking in explicit module build

cachemeifyoucan commented:
Rebase to rebranch

cachemeifyoucan committed:
Virtualize depscan output to allow deterministic check

Batch mode depscan also gets diagnostics for output errors.

cachemeifyoucan committed:
Capture clang importer output in virtual output

cachemeifyoucan committed:
Virtualize swift abi description file

cachemeifyoucan committed:
Handle dependencies output

Support makefile style dependency file and imported modules. Need
improvements to virtual backend to allow emit loaded module trace file.

cachemeifyoucan committed:
Add test cases for virtualized output types

Add unit-tests for already virtualized outputs

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
This PR attempt to virtualized outputs from swift-frontend, which allows operations on the outputs. The example in the PR is to run the compiler twice and check the output to see if output is deterministic. This enables adding caching to the swift compiler because we can redirect the output to CAS.

PR doesn't handle all the output from swift-frontend yet but it handles most of the important ones and can be used as an example to virtualize other output kinds.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

artemcm reviewed (commented):
I don't think I have a better idea on passing around the `VirtualBackend`s in most use-cases here. Perhaps `ClangImporter` can own a reference to one instead of changing its APIs. 

But otherwise it seems necessary. I suppose that if new code gets added that forgets to use a backend for writing some new output, we'd learn pretty quickly that we have a bad CAS citizen.

cachemeifyoucan commented on include/swift/AST/FileSystem.h:
```diff
@@ -16,29 +16,38 @@
 /// operations or from \p action returning true.
 inline bool
-withOutputFile(DiagnosticEngine &diags, StringRef outputPath,
+withOutputFile(DiagnosticEngine &diags, llvm::vfs::OutputBackend &Backend,
```
I was renaming it to `withOutputBackend` but Ben doesn't like the name. Maybe `withOutputPath` is better.

cachemeifyoucan commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -517,8 +529,18 @@ static bool dumpAST(CompilerInstance &Instance) {
       auto OutputFilename = PSPs.OutputFilename;
-      auto OS = getFileOutputStream(OutputFilename, Instance.getASTContext());
-      sourceFile->dump(*OS, /*parseIfNeeded*/ true);
+      auto OutFile = Instance.getOutputBackend().createFile(OutputFilename);
```
Switch to `withOutputPath`

cachemeifyoucan commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -123,37 +130,36 @@ emitLoadedModuleTraceForAllPrimariesIfNeeded(ModuleDecl *mainModule,
+    return true;
+  }
+  SM.print(*OutFile, M, Opts);
+  if (auto E = OutFile->keep()) {
```
There is an assertion failure if you forgot to `keep()` or `discard()` the file. This is generally a good practice for all the outputs since `keep` can certainly error depending on the backend or doing an atomic rename of the file.

cachemeifyoucan commented on include/swift/ClangImporter/ClangImporter.h:
```diff
@@ -394,8 +395,9 @@ class ClangImporter final : public ClangModuleLoader {
-  bool emitBridgingPCH(StringRef headerPath,
-                       StringRef outputPCHPath);
+  bool
+  emitBridgingPCH(llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> backend,
```
Sure. It depends if we need the flexibility to use a different output backend at call site. It seems that we don't and clang importer already hold ASTContext, which now holds a output backend.

cachemeifyoucan committed:
Address review feedback

cachemeifyoucan commented:
Address review feedback. Yes, for the output that doesn't write into output backend, they will not be stored and replayed in the cache. I only convert most of the common output file in this commit. We can add more in the future as we need to support more.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

benlangmuir commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -1135,6 +1135,13 @@ def enable_emit_generic_class_ro_t_list :
 
+def enable_swift_deterministic_check :
+  Flag<["-"], "enable-swift-deterministic-check">,
+  HelpText<"Check swift compiler output determinisim by run it twice">;
```
This comment wasn't addressed

benlangmuir reviewed (commented):
There were a couple of my minor comments from January that seemed to have been lost, but otherwise LGTM.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test

cachemeifyoucan committed:
Fix abi-digester for non-macos platforms

Fix broken test cases on non-macos platforms.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/apple/llvm-project/pull/6112

@swift-ci please smoke test
